src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/SourceItemHandler.cs

```diff
@@ -0,0 +1,163 @@
+        {
+            Requires.NotNull(project, nameof(project));
+            Requires.NotNull(projectTree, nameof(projectTree));
+
+            _project = project;
+            _projectTree = projectTree;
+        }
+
+        public string RuleName
+        {
+            get { return CSharp.SchemaName; }
+        }
+
+        public RuleHandlerType HandlerType
+        {
+            get { return RuleHandlerType.Evaluation; }
+        }
+
+        public void SetContext(IWorkspaceProjectContext context)
+        {
+            Requires.NotNull(context, nameof(context));
+
+            _context = context;
+        }
+
+        public void Handle(CommandLineArguments added, CommandLineArguments removed)
+        {
+            Requires.NotNull(added, nameof(added));
+            Requires.NotNull(removed, nameof(removed));
+
+            foreach (CommandLineSourceFile sourceFile in removed.SourceFiles)
+            {
+                RemoveSourceFile(sourceFile.Path);
+            }
+
+            foreach (CommandLineSourceFile sourceFile in added.SourceFiles)
+            {
+                AddSourceFile(null,  sourceFile.Path);
+            }
+        }
+
+        public async Task HandleAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> e, IProjectChangeDescription projectChange)
+        {
+            Requires.NotNull(e, nameof(e));
+            Requires.NotNull(projectChange, nameof(projectChange));
+
+            IProjectChangeDiff diff = projectChange.Difference;
+
+            foreach (string filePath in diff.RemovedItems)
+            {
+                // Item includes are always relative to csproj/vbproj
+                string fullPath = _project.MakeRooted(filePath);
+
+                RemoveSourceFile(fullPath);
+            }
+
+            if (diff.AddedItems.Count > 0 || diff.RenamedItems.Count > 0)
+            {
+                // Make sure the tree matches the same version of the evalutation that we're handling
+                IProjectTreeServiceState treeState = await _projectTree.TreeService.PublishTreeAsync(e.ToRequirements())
+                                                                                   .ConfigureAwait(false);
+
+                foreach (string filePath in diff.AddedItems)
+                {
+                    string fullPath = _project.MakeRooted(filePath);
+
+                    AddSourceFile(treeState, fullPath);
+                }
+
+                foreach (KeyValuePair<string, string> filePaths in diff.RenamedItems)
+                {
+                    string beforeFullPath = _project.MakeRooted(filePaths.Key);
+                    string afterFullPath = _project.MakeRooted(filePaths.Value);
+
+                    RemoveSourceFile(beforeFullPath);
+                    AddSourceFile(treeState, afterFullPath);
+                }
+            }
+        }
+
+        private void RemoveSourceFile(string fullPath)
+        {
+            if (_sourceFiles.Remove(fullPath))
```

Comment:
This is the part of the code review where I ask is it possible for Handle or HandleAsync to be called multiple times at the same time, introducing a race